### PR TITLE
Add history to the search bar

### PR DESCRIPTION
**Pull request by MathieuxC**
This PR add a history to the search bar. This means that when starting a research and until the user start typing something, he will see the last 5 research he made. If he hadn't made any, the screen still looks like before. There is no way for now to remove a user from history or to see the whole list of all research.

### CODE REVIEW

File: app/src/main/java/com/github/se/eduverse/repository/ProfileRepository.kt [hunk 1]
```diff
@@ -214,6 +220,39 @@ open class ProfileRepositoryImpl(
...
+  }
+
+  override suspend fun addProfileToHistory(userId: String, searchedProfileId: String) {
+    usersCollection
+        .document(userId)
+        .collection("searchHistory")
+        .add(hashMapOf("timestamp" to System.currentTimeMillis(), "id" to searchedProfileId))
+        .await()
+  }
+
```
Comment (Fayroth) [inline app/src/main/java/com/github/se/eduverse/repository/ProfileRepository.kt:255]: The implementation of the addProfileToHistory method is straightforward and leverages Firestore's ability to add documents to a collection with a timestamp.

File: app/src/main/java/com/github/se/eduverse/repository/ProfileRepository.kt [hunk 2]
```diff
@@ -214,6 +220,39 @@ open class ProfileRepositoryImpl(
...
+          .whereIn(FieldPath.documentId(), listId)
+          .get()
+          .await()
+          .documents
+          .mapNotNull { it.toObject(Profile::class.java) }
+    } catch (e: Exception) {
+      Log.e("LOAD_HISTORY", "Failed to load search history: ${e.message}")
+      emptyList()
+    }
+  }
```
Comment (Fayroth) [inline app/src/main/java/com/github/se/eduverse/repository/ProfileRepository.kt:246]: The handling of exceptions in loadSearchHistory improves the robustness of the code by catching potential errors and logging them. well done

File: app/src/main/java/com/github/se/eduverse/repository/ProfileRepository.kt
Comment (Fayroth) [inline app/src/main/java/com/github/se/eduverse/repository/ProfileRepository.kt:1]: Overall, the additions are well-implemented and follow good practices for handling asynchronous operations with Firestore. If the search history grows significantly, fetching the full list might impact performance. Implementing pagination or a time-based cleanup mechanism could help manage this. maybe think about this for a future small PR that handles this but for now its good.

File: app/src/main/java/com/github/se/eduverse/ui/search/SearchProfileScreen.kt
Comment (Fayroth) [inline app/src/main/java/com/github/se/eduverse/ui/search/SearchProfileScreen.kt:1]: LGTM, the inclusion of the currentUserId parameter in the SearchProfileScreen function makes it easier to track the user's search history. and for ux purposes consider implementing a loading indicator for the search history to inform users when the history is being fetched, improving the user experience. You can look at how its done in the profile viewmodel and the profile screen it's quite easy to do imo.

File: app/src/test/java/com/github/se/eduverse/repository/ProfileRepositoryTest.kt
Comment (Fayroth) [inline app/src/test/java/com/github/se/eduverse/repository/ProfileRepositoryTest.kt:1]: Your tests are complete and well implemented good job

File: app/src/test/java/com/github/se/eduverse/viewmodel/ProfileViewModelTest.kt
Comment (Fayroth) [inline app/src/test/java/com/github/se/eduverse/viewmodel/ProfileViewModelTest.kt:1]: good check for errors and normal behavior good job

File: app/src/main/java/com/github/se/eduverse/viewmodel/ProfileViewModel.kt
Comment (Fayroth) [inline app/src/main/java/com/github/se/eduverse/viewmodel/ProfileViewModel.kt:1]: As i said for the repo its all good, the changes are well-structured and improve the application's functionality, following good practices for state management and asynchronous operations. the only suggestion i can make is the optimization part that could have a look to see if it's easy to implement or not.

**Review by Fayroth — CHANGES_REQUESTED** [review]
## Review Summary - Search History Implementation

This PR introduces a user-friendly search history feature, enabling users to view their recent searches. The implementation is clean, well-tested, and maintains good practices throughout.

### Key Strengths

- **Feature Implementation**
 - Clean integration of search history with existing search functionality
 - Smart limitation to last 5 searches for optimal UX
 - Good use of Firestore for history storage
 - Proper timestamp tracking for chronological order

- **Code Quality**
 - Well-structured exception handling
 - Clean repository and ViewModel integration
 - Good separation of concerns
 - Robust error logging

- **Testing**
 - Comprehensive test coverage for both repository and ViewModel
 - Good validation of error scenarios
 - Clear test organization
 - Thorough edge case handling

### Required Changes

1. Add loading state handling during history fetch:
  - Implement loading indicator in SearchProfileScreen similar to ProfileScreen implementation
  - This is a relatively quick addition that will significantly improve user experience

### Suggested Future Enhancements
Consider for future iterations:
- Implement pagination or cleanup for large history lists
- Enable history item deletion
- Add full history view capability

### Verdict

**Changes Requested**

The implementation is solid overall with good testing coverage. However, adding a loading indicator for the history fetch would provide better user feedback and align with the app's UX patterns. This is a small change that shouldn't take long to implement.

Once the loading indicator is added, this PR will be ready for approval. Great work on the core functionality!

File: app/src/main/java/com/github/se/eduverse/ui/search/SearchProfileScreen.kt
Comment (MathieuxC) [reply app/src/main/java/com/github/se/eduverse/ui/search/SearchProfileScreen.kt:1]: There is actualy already a loading indicator in the ui (it was already implemented in SearchProfileScreen when the searchState value of the viewmodel is Loading). I added the required change in said parameter (viewmodel l.194) so that it displays correctly. See screen recording below (it goes pretty fast but there is a loading indicator during the loading time)

https://github.com/user-attachments/assets/e7b1df97-f254-4f8e-95eb-11d9ea4caddc

**Review by Fayroth — APPROVED** [review]

**Comment by Fayroth** [comment]
Everything is fine then ive approved the changes
